pkg/metrics/prometheus.go

```diff
@@ -0,0 +1,152 @@
+package metrics
+
+import (
+	"github.com/grafana/grafana/pkg/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"net/http"
+	"strings"
+)
+
+const (
+	namespace = "grafana"
+	subsystem = "backend"
+)
+
+type PrometheusPublisher struct {
+	CounterCollectors map[string]*prometheus.CounterVec
+	GaugeCollectors   map[string]*prometheus.GaugeVec
+}
+
+func CreatePrometheusPublisher() (*PrometheusPublisher, error) {
+	go func() {
+		http.Handle("/metrics", prometheus.Handler())
+		err := http.ListenAndServe(":9200", nil)
```

Comment:
> That's not sufficient, Prometheus isn't designed to work with data like this. There's a difference between data being in the Prometheus format and the data being useful for Prometheus.

So prometheus needs its golang client library embedded in all servers/apps it wants to monitor? 

>A lot of things are rates. For example in the given output, there's only 2 pieces of data being exposed which are fundamentally gauges. The other ~20 are various forms of counter.

Yes, and storing counters as deltas makes analytics & rollups so much easier and powerful. 

> These have the same issues. We expect those to be sent as two counters.

You store timers as two counts? The timers in Grafana provide these metrics:
```
"api.dashboard.search": {
  "count": 3,
  "max": 0,
  "mean": 0,
  "min": 0,
  "p25": 0,
  "p75": 0,
  "p90": 0,
  "p99": 0,
  "std": 0
 },
```

On the question of storing/sending counter deltas vs raw value. It has its pros and cons. Sending a counter that never resets is more resilient to packet loss, etc. But storing them as that is to me very bad in my experience as I am usually coming to metrics, not from a pure monitoring perspective but an analytics perspective wanting to know the short and long term trending of application & user behavior, working with rate functions that you then have to multiply to get bigger time periods or sum is very inaccurate or complex and usually breaks down when you hit roll ups periods (not a problem for Prometheus I guess as it does not support rollups). Storing deltas makes analytics so much easier, you can just sum them, and the rollups are summed so no matter what time range you can get accurate sums. And super easy to turn them into rates with graphites scaleToSeconds (which knows how many seconds each value represents so it's "accurate" no matter what rollup you hit).  

Anyone, no matter how interesting that discussion is, this issue needs to be solved some way that does not involve adding a new metrics lib.